agent/acs/handler/heartbeat_handler.go

```diff
@@ -0,0 +1,119 @@
+	"github.com/aws/amazon-ecs-agent/agent/wsclient"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/cihub/seelog"
+)
+
+// heartbeatHandler handles heartbeat messages from ACS
+type heartbeatHandler struct {
+	heartbeatMessageBuffer    chan *ecsacs.HeartbeatMessage
+	heartbeatAckMessageBuffer chan *ecsacs.HeartbeatAckRequest
+	ctx                       context.Context
+	cancel                    context.CancelFunc
+	acsClient                 wsclient.ClientServer
+}
+
+// newHeartbeatHandler returns an instance of the heartbeatHandler struct
+func newHeartbeatHandler(ctx context.Context,
+	acsClient wsclient.ClientServer) heartbeatHandler {
+
+	// Create a cancelable context from the parent context
+	derivedContext, cancel := context.WithCancel(ctx)
+	return heartbeatHandler{
+		heartbeatMessageBuffer:    make(chan *ecsacs.HeartbeatMessage),
+		heartbeatAckMessageBuffer: make(chan *ecsacs.HeartbeatAckRequest),
+		ctx:                       derivedContext,
+		cancel:                    cancel,
+		acsClient:                 acsClient,
+	}
+}
+
+// handlerFunc returns a function to enqueue requests onto the buffer
+func (heartbeatHandler *heartbeatHandler) handlerFunc() func(message *ecsacs.HeartbeatMessage) {
+	return func(message *ecsacs.HeartbeatMessage) {
+		heartbeatHandler.heartbeatMessageBuffer <- message
+	}
+}
+
+// start() invokes go routines to handle receive and respond to heartbeats
+func (heartbeatHandler *heartbeatHandler) start() {
+	go heartbeatHandler.handleHeartbeatMessage()
+	go heartbeatHandler.sendHeartbeatAck()
+}
+
+func (heartbeatHandler *heartbeatHandler) handleHeartbeatMessage() {
+	for {
+		select {
+		case message := <-heartbeatHandler.heartbeatMessageBuffer:
+			if err := heartbeatHandler.handleSingleHeartbeatMessage(message); err != nil {
+				seelog.Warnf("Unable to handle heartbeat message [%s]: %v", message.String(), err)
+			}
+		case <-heartbeatHandler.ctx.Done():
+			return
+		}
+	}
+}
+
+func (heartbeatHandler *heartbeatHandler) handleSingleHeartbeatMessage(message *ecsacs.HeartbeatMessage) error {
+	seelog.Tracef("Received server heartbeat message: %s", message.MessageId)
+	go func() {
+		response := &ecsacs.HeartbeatAckRequest{
+			MessageId: message.MessageId,
+		}
+		heartbeatHandler.heartbeatAckMessageBuffer <- response
+	}()
+	return nil
+}
+
+func (heartbeatHandler *heartbeatHandler) sendHeartbeatAck() {
+	for {
+		select {
+		case ack := <-heartbeatHandler.heartbeatAckMessageBuffer:
+			heartbeatHandler.sendSingleHeartbeatAck(ack)
+		case <-heartbeatHandler.ctx.Done():
+			return
+		}
+	}
+}
+
+func (heartbeatHandler *heartbeatHandler) sendSingleHeartbeatAck(ack *ecsacs.HeartbeatAckRequest) {
+	err := heartbeatHandler.acsClient.MakeRequest(ack)
+	if err != nil {
+		seelog.Warnf("Error acknowledging server heartbeat, message id: %s, error: %v", aws.StringValue(ack.MessageId), err)
+	}
+	seelog.Tracef("Acknowledging server heartbeat message: %s", ack.MessageId)
```

Comment:
This will print a bit funny since MessageId is a pointer, but in the trace/debug messages we could probably just stringify the whole object anyways, ie just print `ack.String()`